sprint 1/step 1

mellonis opened the pull request.

mellonis committed:
Build app structure

- Adds App, AppHeader and AppBody components
- Adds BurgerIngredients and BurgerConstructor stubs

mellonis committed:
Adds temporary data

mellonis committed:
Apply recommendations from the task

mellonis committed:
Some refactoring

- Adds MenuItem component
- Adds prettier config file

mellonis committed:
Adds BurgerIngredients layout

mellonis committed:
BurgerIngredients. Returns render of all ingredients

mellonis committed:
BurgerIngredients. Fixes render errors

mellonis committed:
BurgerConstructor. Adds component implementation

- Adds the Amount component
- Some refactoring

gennady-bars commented on src/components/app-body/index.tsx:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
```
**Можно лучше**

в каждой папке `index.tsx` используется в программировании, но это очень сбивает с толку, если открыть несколько файлов и у Вас будут все `index.tsx` во вкладках и можно запутаться.

Лучше называть файл так, как сама папка и экспортировать из него.


```
app-body.tsx

```

Вот скрин того, как это выглядит сейчас http://joxi.ru/DmBYyBDS4R3E7r

mellonis replied:
А при импорте нужно будет писать import Component from '../components/component/component'?

Или loader поймёт что если указана папка, загружать либо index.tsx или одноименный с ней tsx-файл?

gennady-bars replied:
> А при импорте нужно будет писать import Component from '../components/component/component'?
> 
> Или loader поймёт что если указана папка, загружать либо index.tsx или одноименный с ней tsx-файл?

 придется дублировать имя папки и имя файла. Но чаще всего есть автоимпорт в IDE (в средах разработки) и не нужно думать об импортах при этом

gennady-bars commented on src/components/app-body/index.tsx:
```diff
@@ -0,0 +1,22 @@
+        'pl-5 pr-5 text text_type_main-default'
+      )}
+    >
+      <BurgerIngredients className={styles['app-body__ingredients']} />
```
`className={styles['app-body__ingredients']}`

эти стили можно было поместить внутри компонентов, а не передавать через `props`

mellonis replied:
Здесь следовал BEM, конкретно этот класс описывает элемент блока ingredients блока app-body.

Компонент BurgerIngredients не знает, что должен выводится именно в этом блоке, поэтому считаю, что лучше передавать через пропс. Как думаете?

gennady-bars replied:
> Здесь следовал BEM, конкретно этот класс описывает элемент блока ingredients блока app-body.
> 
> Компонент BurgerIngredients не знает, что должен выводится именно в этом блоке, поэтому считаю, что лучше передавать через пропс. Как думаете?

Если по BEM, то можно и оставить. Хотя `BurgerIngredients ` у Вас только один в проекте, поэтому можно в его `css` так и указать с этим классом, который никогда не изменится

gennady-bars commented on src/components/app-header/index.tsx:
```diff
@@ -0,0 +1,43 @@
+    <header className={cs(styles['app-header'], 'text text_type_main-default')}>
+      <nav>
+        <ul className={cs(styles['app-header__menu-list'], 'pt-4 pb-4')}>
+          <MenuItem
```
нужно еще добавить в стили, чтобы отображалась не стрелочка в курсоре, а `рука`, которая показывает, что можно кликнуть на элемент

mellonis replied:
Готово

gennady-bars commented on src/components/burger-constructor/index.tsx:
```diff
@@ -0,0 +1,156 @@
+};
+
+const BurgerConstructor = ({ className }: { className?: string }) => {
+  const [{ list, total }, dispatch] = useReducer(
```
`useReducer`

даже подсчет стоимости сделали с `useReducer`. Супер

gennady-bars commented on src/components/burger-ingredients/burger-ingredient-type/index.tsx:
```diff
@@ -0,0 +1,39 @@
+          <React.Fragment key={ix}>
+            <BurgerIngredient data={ingredient} />
+            <li
+              key={`${ix}:g`}
```
`  key={`${ix}:g`}`

атрибут `key` нужен только на самом верхнем уровне вложенности (в обертке верстки), а тут он уже не нужен, так как оберткой является ` <React.Fragment key={ix}>` на 23й строчке

mellonis replied:
У BurgerIngredient убрал, у второго элемента забыл. Поправлю.

mellonis replied:
Готово

gennady-bars reviewed (commented):
Здравствуйте. Работа проделана огромная, но можно сделать чуть лучше:

* атрибут `key` нужен только на самом верхнем уровне вложенности (в обертке верстки)
* в каждой папке `index.tsx` используется в программировании, но это очень сбивает с толку, если открыть несколько файлов и у Вас будут все `index.tsx` во вкладках и можно запутаться.

 Пожалуйста, проверьте работоспособность проекта перед отправкой на ревью.

Удачного рефакторинга кода.

mellonis commented:
@gennady-bars , есть ещё общий вопрос. В задании указано, что отступы нужно брать из UI библиотеки. Сейчас это можно сделать только используя классы (pt-1, mt-2 и др.) Из-за чего мне приходится создавать, много служебных элементов, наличие которых просто занимать место между элементами.

Так по большей части я бы смог это делать с помощью grid-gap непосредственно в css, но вынужден создавать служебные элементы с классами, регулирующими отступы.

Это ничего в нашем проекте?

mellonis committed:
MenuItem. Adds pinter cursor for an actice menu-item

mellonis committed:
BurgerIngredientType. Removes unneeded key prop

mellonis committed:
Refactors

- Apply border-box box-sizing to all elements
- BurgerIngredients, BurgerConstructor. Adds scrollbar and customize their styles for

gennady-bars commented:
> @gennady-bars , есть ещё общий вопрос. В задании указано, что отступы нужно брать из UI библиотеки. Сейчас это можно сделать только используя классы (pt-1, mt-2 и др.) Из-за чего мне приходится создавать, много служебных элементов, наличие которых просто занимать место между элементами.
> 
> Так по большей части я бы смог это делать с помощью grid-gap непосредственно в css, но вынужден создавать служебные элементы с классами, регулирующими отступы.
> 
> Это ничего в нашем проекте?

Ничего страшного. Если слишком нелогично использовать классы из библиотеки, то можно свои `css` стили сделать.

gennady-bars commented on src/components/burger-ingredients/burger-ingredient-type/index.tsx:
```diff
@@ -23,7 +23,6 @@ const BurgerIngredientType = ({
           <React.Fragment key={ix}>
             <BurgerIngredient data={ingredient} />
             <li
-              key={`${ix}:g`}
```
теперь все отлично

gennady-bars commented on src/components/app-body/style.module.css:
```diff
@@ -3,11 +3,32 @@
+}
 
+.app-body__constructor {
+    overflow-y: scroll;
```
Нужно посмотреть по макету в фигме, где стролл находится. А то сейчас весь бургер прокручивается, а нужно прокручивать ингредиенты внутри булок, если их больше, чем помещается в высоту экрана.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
